### PR TITLE
Keep QuantityPoint conversion within its Rep

### LINKED ISSUE #100: QuantityPoint::InheritsOverflowSafetySurfaceFromUnderlyingQuantityTypes test broken
This is one of those "uncomment to test" cases, because it is supposed to produce a hard compiler error.  I uncommented it, and it didn't produce the error.

I think there are two contributing causes.

1. We respect the underlying integer promotion rules of C++ (inherited from C).  This means that the subtraction that takes place inside of the `QuantityPoint` conversion gets promoted to `int` when it takes place between two 16-bit types.

2. Our origin happens to be defined with Rep `int`, which means the subtraction with a 16-bit Rep would produce `int` even if there weren't any integer promotion.

Impact
------

I don't see any safety risks here.  It looks like just a small annoyance for people who are trying to use smaller integral Reps, _and_ converting temperatures from one scale to another.

Remedy
------

The minimal remedy is to fix up the test case.  I'll see if I can tackle the underlying behaviour as well.

### PR DESCRIPTION
**Pull request by chiphogg**
If we call `celsius_pt(x).in(kelvins_pt / mag<20>())`, the result should
have the same type as `x`.  Currently, that's not the case for two
reasons, both of which stem from the underlying operation of
_subtracting the origin displacement_.

1. The origin itself typically has `int` Rep.

2. Even if it didn't, C++'s integer promotion rules would take us
   outside the realm of shorter integral types.

Part of the solution is to cast the origin displacement Rep to the
desired Rep type explicitly.  There's an opportunity for a safety check
here, so we add it (including unit tests).

The rest of the solution is simply to `rep_cast` at the end to "undo"
any integer promotion that may have happened.

Along the way, we also changed the operation from _subtraction_ of one
origin displacement, to _addition_ of the inverse displacement.  Now
that we're being more careful about checking whether the origin
displacement fits in the Rep, this would have broken unsigned type
support if we hadn't made this change.  (What I think was happening
before with the unsigned types was probably promotion to signed integers
--- so, no "incorrect" results, but probably an unexpected Rep.)

We also needed to add a trivial `rep_cast` overload for `Zero`.  This
helps it "stand in" for arbitrary quantities in one more way.

Fixes #100.

Test plan
---------

- [x] Added new test cases for `OriginDisplacementFitsIn`.
- [x] Uncommented the test case broken in #100, and verified that it
      fails for the right reason.
- [x] Test that there is no unacceptable compile time penalty.

### CODE REVIEW

File: au/quantity_point_test.cc [hunk 1]
```diff
@@ -33,7 +33,7 @@ constexpr QuantityMaker<Kelvins> kelvins{};
 constexpr QuantityPointMaker<Kelvins> kelvins_pt{};
 
 struct Celsius : Kelvins {
-    static constexpr auto origin() { return centi(kelvins)(273'15); }
+    static constexpr auto origin() { return (kelvins / mag<20>())(273'15 / 5); }
```
Comment (geoffviola) [inline au/quantity_point_test.cc:36]: Can you comment on the `mag<20>()` and `/ 5` parts?

File: au/quantity_point_test.cc [hunk 2]
```diff
@@ -304,4 +304,34 @@ TEST(QuantityPointMaker, CanScaleByMagnitude) {
     StaticAssertTypeEq<decltype(kelvins_pt / mag<5>()),
                        QuantityPointMaker<decltype(Kelvins{} / mag<5>())>>();
 }
+
+namespace detail {
+TEST(OriginDisplacementFitsIn, CanRetrieveValueInGivenRep) {
+    EXPECT_TRUE((OriginDisplacementFitsIn<uint64_t, Kelvins, Celsius>::value));
```
Comment (geoffviola) [inline au/quantity_point_test.cc:310]: Does it make sense to test a case where the origin displacent results in a negative number and doesn't fit in the unsigned value?

File: au/quantity_point_test.cc [hunk 3]
```diff
@@ -270,7 +270,7 @@ TEST(QuantityPoint, CanCompareUnitsWithDifferentOrigins) {
 }
 
 TEST(QuantityPoint, CanSubtractIntegralInputsWithNonintegralOriginDifference) {
-    EXPECT_THAT(celsius_pt(0) - kelvins_pt(273), QuantityEquivalent(centi(kelvins)(15)));
+    EXPECT_EQ(celsius_pt(0) - kelvins_pt(273), centi(kelvins)(15));
```
Comment (geoffviola) [inline au/quantity_point_test.cc:273]: Commend: Less casting. Looks good!

**Review by geoffviola — APPROVED** [review]
LGTM

In the summary, it says

> If we call celsius_pt(x).in(kelvins), the result should have the same
type as x

I get that the intention on the output rep, but I couldn't get that code to work with `kelvins` or with integers and `kelvins_pt`.

```
celsius_pt(0).in(kelvins)
```
There's a compiler error about no matching member function for call to 'in'.

```
celsius_pt(0).in(kelvins_pt);
```
there's `static_assert Dangerous conversion`.

It seems like it runs the same on main so it doesn't seem too concerning.


Also, I ran this test
```
celsius_pt(static_cast<uint8_t>(0)) - kelvins_pt(static_cast<uint8_t>(0));
```
On main, it compiles. In this PR, there's `static_assert Dangerous conversion`.

**Review by chiphogg — COMMENTED** [review]
> In the summary, it says
> 
> > If we call celsius_pt(x).in(kelvins), the result should have the same
> > type as x
> 
> I get that the intention on the output rep, but I couldn't get that code to work with `kelvins` or with integers and `kelvins_pt`.
> 
> ```
> celsius_pt(0).in(kelvins)
> ```
> 
> There's a compiler error about no matching member function for call to 'in'.
> 
> ```
> celsius_pt(0).in(kelvins_pt);
> ```
> 
> there's `static_assert Dangerous conversion`.
> 
> It seems like it runs the same on main so it doesn't seem too concerning.

Oops --- I was too hasty in writing that in the PR summary.  Clearly, I never tried compiling it!  There are two errors.

1. I should have written `kelvins_pt`, instead of `kelvins`.
2. I needed to divide by (at least) 20, because otherwise this will of course fail for every integral Rep.

I plan to add a test for this, but we'll need to tweak the overflow surface threshold first, which I'm planning to do as part of fixing #69 anyway.

> Also, I ran this test
> 
> ```
> celsius_pt(static_cast<uint8_t>(0)) - kelvins_pt(static_cast<uint8_t>(0));
> ```
> 
> On main, it compiles. In this PR, there's `static_assert Dangerous conversion`.

Perfect!
